File: VBNetCore.Mvvm/Helpers/PopupManager.vb
Imports Microsoft.Extensions.DependencyInjection
Imports VBNetCore.Mvvm.Abstractions
Imports VBNetCore.Mvvm.Controls
Imports VBNetCore.Mvvm.ViewModels
Imports System

Namespace VBNetCore.Mvvm.Helpers
    Public Class PopupManager
        Implements IPopupManager
        Private ReadOnly _serviceProvider As IServiceProvider
        Private _popupMessageViewModel As PopupMessageViewModel
        Public Sub New(ByVal serviceProvider As IServiceProvider, ByVal popupMessageViewModel As PopupMessageViewModel)
            _serviceProvider = serviceProvider
            _popupMessageViewModel = If(popupMessageViewModel, _serviceProvider.GetRequiredService(Of PopupMessageViewModel)())
        End Sub

        Public Function ShowPopup(Of T As PopupViewModel)(ByVal popupViewModel As T, ByVal owner As IPopupable) As T Implements IPopupManager.ShowPopup
            If owner IsNot Nothing AndAlso TypeOf owner Is IPopupable Then
                Return TryCast(owner.ShowPoup(popupViewModel), T)
            End If
            Return Nothing
        End Function

        Public Function ShowPopupMessage(ByVal message As String, ByVal caption As String, ByVal owner As IPopupable, ByVal Optional popupButton As PopupButton = PopupButton.OK, ByVal Optional popupImage As PopupImage = PopupImage.None) As PopupResult Implements IPopupManager.ShowPopupMessage
            If owner IsNot Nothing AndAlso TypeOf owner Is IPopupable Then
                _popupMessageViewModel.Message = message
                _popupMessageViewModel.Caption = caption
                _popupMessageViewModel.Button = popupButton
                _popupMessageViewModel.Image = popupImage

                Dim popup = owner.ShowPoup(_popupMessageViewModel)
                If popup IsNot Nothing Then Return popup.Result

            End If
            Return PopupResult.None
        End Function
    End Class
End Namespace

File: VBNetCore.Mvvm/Helpers/PopupManager.cs
using Microsoft.Extensions.DependencyInjection;
using NetCore.Mvvm.Abstractions;
using NetCore.Mvvm.Controls;
using NetCore.Mvvm.ViewModels;
using System;
using System.Linq;

namespace NetCore.Mvvm.Helpers
{
    public class PopupManager : IPopupManager
    {
        private readonly IServiceProvider _serviceProvider;
        private PopupMessageViewModel _popupMessageViewModel;
        public PopupManager(IServiceProvider serviceProvider, PopupMessageViewModel popupMessageViewModel)
        {
            _serviceProvider = serviceProvider;
            _popupMessageViewModel = popupMessageViewModel ?? _serviceProvider.GetRequiredService<PopupMessageViewModel>();
        }

        public T? ShowPopup<T>(T popupViewModel, IPopupable owner) where T : PopupViewModel
        {
            if(owner != null && owner is IPopupable)
            {
                return owner.ShowPoup(popupViewModel) as T;
            }
            return null;
        }

        public PopupResult ShowPopupMessage(string message, string caption, IPopupable owner, PopupButton popupButton = PopupButton.OK, PopupImage popupImage = PopupImage.None)
        {
            if (owner != null && owner is IPopupable)
            {
                _popupMessageViewModel.Message = message;
                _popupMessageViewModel.Caption = caption;
                _popupMessageViewModel.Button = popupButton;
                _popupMessageViewModel.Image = popupImage;

                var popup = owner.ShowPoup(_popupMessageViewModel);
                if (popup != null)
                    return popup.Result;

            }
            return PopupResult.None;
        }
    }
}
